
# CS-GY 6903 Applied Cryptography
## Project 1
### Nicola Maiorana (nam10102)
### Spring 2025, NYU Tandon School of Engineering

Submission
1. Project report using PDF format (proj1_your_name.pdf).
    - Ensure that your submission is clear, well-organized, and comprehensively covers all aspects of the assignment.
    - Use appropriate headings and mark the problem numbers clearly for easy reference.
    - May include codes in the report. Students still need to submit the Jupyter Notebook.
2. Jupyter Notebook (proj1-Your_name.ipynb)
    - Students shall use Jupyter Notebook for the coding.
    - The Jupyter Notebook file shall contain all the coding solutions. It should be well-commented, indicating what each segment does, and should correspond to the problems as numbered in the assignment.
    - Ensure that the notebook is executable without errors and that the outputs of your code are visible. For example, add all the pip install statements in a code box at the top of your ipynb notebook if they are not installed already in the CA’s machine.


## Install the required modules

In [5]:
!pip install jupyter==1.1.1
!pip install notebook==7.3.2


## Problem 1: Vigenère cipher

(a) Encrypt the message using the Vigenère cipher with the given key, k = ‘apple’ to the message, m = ‘theclassistough’ to produce the ciphertext. Then (b) decrypt the ciphertext back to retrieve the original plaintext. Show your work.

- Key is of the form K = (k0,k1,...,kn-1), Where each ki E {0,1,2,...,25}
- Encryption:   ci= pi+ ki (mod n)(mod 26), i= 0,1,2, ... , L
- Decryption:   pi= ci–ki (mod n)(mod 26), i= 0,1,2, ... , L
- Just a repeating sequence of (shift by n) simple substitutions•Obviously, when n = 1, it reduces to the Caesar Cipher. 34


### Experimenting with the math

In [15]:
k = 'apple'
p = 'theclassistough'
n = len(p)
print(f'n is {n}')

# key Generation
# Key is of the form K = (k0,k1,...,kn-1), Where each ki E {0,1,2,...,25}
# The generated key ends up being the k value fit into the length of the message as many times as possible
g = 'appleappleapplea'

# Encryption
# Encryption:   ci= pi+ ki (mod n)(mod 26), i= 0,1,2, ... , L
# For the first letter
pi = 't'
ki = 'a'
print(f"'{pi}' is {ord(pi)}")
print(f"'{ki}' is {ord(ki)}")
ki_mod_15 = ord(ki) % n
print(f"{ord(ki)} mod {n} is {ki_mod_15}")
ki_mod_15_mod_26 = ki_mod_15 % 26
print(f'{ki_mod_15} mod 26  is {ki_mod_15_mod_26}')
ci = ord(pi) + ki_mod_15_mod_26
print(f"ci is '{chr(ci)}': {ci} = {ord(pi)} + {ki_mod_15_mod_26}")

# Decryption
# Decryption:   pi= ci–ki (mod n)(mod 26), i= 0,1,2, ... , L
# For the first letter
di = ci - ki_mod_15_mod_26
print(f"di is '{chr(di)}': {di} = {ci} - {ki_mod_15_mod_26}")

n is 15
't' is 116
'a' is 97
97 mod 15 is 7
7 mod 26  is 7
ci is '{': 123 = 116 + 7
di is 't': 116 = 123 - 7


## Key Generaton:
Part of problem 1 a and b is to generate the key which is the original key, repeating as many times as needed to match the length of the message. Since our key is 'apple' and our message is 'theclassistough', our key will looke like this (showing the message above to demonstrate they are the same length):
```plaintext
'theclassistough'
'appleappleapple'
```

## Part a: Encrypt the message using the Vigenère cipher with the given key, k = ‘apple’ to the message, m = ‘theclassistough’ to produce the ciphertext.
For this we need to start by encoding the letters of both the message and they key:

- 'theclassistough'
    - t = 116
    - h = 104
    - e = 101
    - c = 99
    - l = 108
    - a = 97
    - s = 115
    - s = 115
    - i = 105
    - s = 115
    - t = 116
    - o = 111
    - u = 117
    - g = 103
    - h = 104
- 'appleappleapple'
    - a = 97
    - p = 112
    - p = 112
    - l = 108
    - e = 101
    - a = 97
    - p = 112
    - p = 112
    - l = 108
    - e = 101
    - a = 97
    - p = 112
    - p = 112
    - l = 108
    - e = 101

Now that we have the ascii values for each of the characters, we can encrypt the message using the following formula:
- ci= pi+ ki (mod n)(mod 26), i= 0,1,2, ... , L
- n is 15 which is the number of characters in the message
    - '{' or 123 = 116 + ((97 mod 15) mod 26)
    - 'o' or 111 = 104 + ((112 mod 15) mod 26)
    - 'l' or 108 = 101 + ((112 mod 15) mod 26)
    - 'f' or 102 = 99 + ((108 mod 15) mod 26)
    - 'w' or 119 = 108 + ((101 mod 15) mod 26)
    - 'h' or 104 = 97 + ((97 mod 15) mod 26)
    - 'z' or 122 = 115 + ((112 mod 15) mod 26)
    - 'z' or 122 = 115 + ((112 mod 15) mod 26)
    - 'l' or 108 = 105 + ((108 mod 15) mod 26)
    - '~' or 126 = 115 + ((101 mod 15) mod 26)
    - '{' or 123 = 116 + ((97 mod 15) mod 26)
    - 'v' or 118 = 111 + ((112 mod 15) mod 26)
    - '|' or 124 = 117 + ((112 mod 15) mod 26)
    - 'j' or 106 = 103 + ((108 mod 15) mod 26)
    - 's' or 115 = 104 + ((101 mod 15) mod 26)
- And the ciphertext is: '{olfwhzzl~{v|js'

## Part b: Decrypt the ciphertext back to retrieve the original plaintext. Show your work.
- Decryption:   pi= ci–ki (mod n)(mod 26), i= 0,1,2, ... , L
- n is 15 which is the number of characters in the ciphertext
    - 't' or 116 = 123 - ((97 mod 15) mod 26)
    - 'h' or 104 = 111 - ((112 mod 15) mod 26)
    - 'e' or 101 = 108 - ((112 mod 15) mod 26)
    - 'c' or 99 = 102 - ((108 mod 15) mod 26)
    - 'l' or 108 = 119 - ((101 mod 15) mod 26)
    - 'a' or 97 = 104 - ((97 mod 15) mod 26)
    - 's' or 115 = 122 - ((112 mod 15) mod 26)
    - 's' or 115 = 122 - ((112 mod 15) mod 26)
    - 'i' or 105 = 108 - ((108 mod 15) mod 26)
    - 's' or 115 = 126 - ((101 mod 15) mod 26)
    - 't' or 116 = 123 - ((97 mod 15) mod 26)
    - 'o' or 111 = 118 - ((112 mod 15) mod 26)
    - 'u' or 117 = 124 - ((112 mod 15) mod 26)
    - 'g' or 103 = 106 - ((108 mod 15) mod 26)
    - 'h' or 104 = 115 - ((101 mod 15) mod 26)
- And the plaintext is: 'theclassistough'

## Problem2: Affine Cipher
The Affine cipher is a type of monoalphabetic substitution cipher that employs modular arithmetic to encrypt the letters of a message. The encryption formula is c = (a * p + b) mod n, where 'p' is the plaintext letter, 'c' is the ciphertext letter, 'n' is the modulus. 'a' and 'b' are keys, and 'a' must be chosen such that 'a' and 'n' are coprime (relatively prime). In this cipher, each letter of the plaintext is converted into an integer, for example, A=0, B=1, C=2, and so on.

a) What is the size of key space for a fixed modular n? Students can use the notation of the Euler's totient Φ(n) which is defined as the number of integers less than n and relative prime to n.